File: frontend/src/utils/tradeValidation.ts
import { CreateTradeRequest, TradeDirection } from "../types/trade";
import {
  FillDraftRow,
  FillSummary,
  buildFillWritePayload,
  computeFillSummary,
  getFillValidationErrors,
} from "./fills";

export interface TradeDraftInput {
  date: string;
  ticker: string;
  direction: TradeDirection;
  entryPrice: string;
  exitPrice: string;
  quantity: string;
  setupId: string;
  emotionId: string;
  ruleFollowed: boolean;
  notes: string;
  useFills: boolean;
  fillRows: FillDraftRow[];
}

export interface TradeDraftValidationResult {
  errors: string[];
  payload: CreateTradeRequest | null;
  fillSummary: FillSummary | null;
}

const VALID_DIRECTIONS: TradeDirection[] = ["CALL", "PUT"];
const CONTRACT_MULTIPLIER = 100;

function isValidDateString(value: string): boolean {
  if (!/^\d{4}-\d{2}-\d{2}$/.test(value)) {
    return false;
  }

  const date = new Date(`${value}T00:00:00`);
  return Number.isFinite(date.getTime());
}

function parsePositiveInteger(value: string): number | null {
  const parsed = Number(value);
  if (!Number.isInteger(parsed) || parsed <= 0) {
    return null;
  }
  return parsed;
}

function parsePositiveFloat(value: string): number | null {
  const parsed = Number(value);
  if (!Number.isFinite(parsed) || parsed <= 0) {
    return null;
  }
  return parsed;
}

export function validateTradeDraft(input: TradeDraftInput): TradeDraftValidationResult {
  const errors: string[] = [];
  const normalizedTicker = input.ticker.trim().toUpperCase();

  if (!input.date) {
    errors.push("Date is required.");
  } else if (!isValidDateString(input.date)) {
    errors.push("Date must be valid.");
  }

  if (!normalizedTicker) {
    errors.push("Ticker is required.");
  }

  if (!VALID_DIRECTIONS.includes(input.direction)) {
    errors.push("Direction is required.");
  }

  const parsedSetupId = input.setupId ? Number(input.setupId) : null;
  const setupId =
    parsedSetupId !== null && Number.isInteger(parsedSetupId) && parsedSetupId > 0
      ? parsedSetupId
      : null;
  if (input.setupId && setupId === null) {
    errors.push("Setup must be valid when provided.");
  }

  const parsedEmotionId = input.emotionId ? Number(input.emotionId) : null;
  const emotionId =
    parsedEmotionId !== null && Number.isInteger(parsedEmotionId) && parsedEmotionId > 0
      ? parsedEmotionId
      : null;
  if (input.emotionId && emotionId === null) {
    errors.push("Emotion must be valid when provided.");
  }

  if (input.useFills) {
    const fillErrors = getFillValidationErrors(input.fillRows);
    if (fillErrors.length > 0) {
      return { errors: [...errors, ...fillErrors], payload: null, fillSummary: null };
    }

    const fillSummary = computeFillSummary(input.fillRows);
    if (
      fillSummary.avgEntryPrice === null ||
      !Number.isFinite(fillSummary.avgEntryPrice) ||
      fillSummary.avgEntryPrice <= 0
    ) {
      errors.push("Avg Entry must be a finite value greater than zero.");
    }
    if (
      fillSummary.avgExitPrice === null ||
      !Number.isFinite(fillSummary.avgExitPrice) ||
      fillSummary.avgExitPrice <= 0
    ) {
      errors.push("Avg Exit must be a finite value greater than zero.");
    }
    if (!Number.isInteger(fillSummary.matchedQty) || fillSummary.matchedQty <= 0) {
      errors.push("Total Quantity must be a whole number greater than zero.");
    }
    if (!Number.isFinite(fillSummary.realizedPnlUsd)) {
      errors.push("Total PnL must be finite.");
    }

    if (errors.length > 0) {
      return { errors, payload: null, fillSummary };
    }

    return {
      errors: [],
      fillSummary,
      payload: {
        date: input.date,
        ticker: normalizedTicker,
        direction: input.direction,
        entry_price: fillSummary.avgEntryPrice as number,
        exit_price: fillSummary.avgExitPrice as number,
        quantity: fillSummary.matchedQty,
        setup_id: setupId,
        emotion_id: emotionId,
        rule_followed: input.ruleFollowed,
        notes: input.notes.trim() ? input.notes.trim() : null,
        use_fills: true,
        fills: input.fillRows.map(buildFillWritePayload),
      },
    };
  }

  const entryPrice = parsePositiveFloat(input.entryPrice);
  if (entryPrice === null) {
    errors.push("Entry price must be greater than zero.");
  }

  const exitPrice = parsePositiveFloat(input.exitPrice);
  if (exitPrice === null) {
    errors.push("Exit price must be greater than zero.");
  }

  const quantity = parsePositiveInteger(input.quantity);
  if (quantity === null) {
    errors.push("Quantity must be a whole number greater than zero.");
  }

  const computedPnl =
    entryPrice !== null && exitPrice !== null ? Number((exitPrice - entryPrice).toFixed(4)) : null;
  if (computedPnl === null || !Number.isFinite(computedPnl)) {
    errors.push("Computed PnL must be finite.");
  }

  const totalPnlUsd =
    computedPnl !== null && quantity !== null ? computedPnl * quantity * CONTRACT_MULTIPLIER : null;
  if (totalPnlUsd === null || !Number.isFinite(totalPnlUsd)) {
    errors.push("Total PnL must be finite.");
  }

  if (errors.length > 0) {
    return { errors, payload: null, fillSummary: null };
  }

  return {
    errors: [],
    fillSummary: null,
    payload: {
      date: input.date,
      ticker: normalizedTicker,
      direction: input.direction,
      entry_price: entryPrice as number,
      exit_price: exitPrice as number,
      quantity: quantity as number,
      setup_id: setupId,
      emotion_id: emotionId,
      rule_followed: input.ruleFollowed,
      notes: input.notes.trim() ? input.notes.trim() : null,
    },
  };
}

File: backend/app/services/trade_validation.py
from __future__ import annotations

from dataclasses import dataclass
from math import isfinite

from fastapi import HTTPException, status

from app.models.trade import Trade
from app.schemas.trade import TradeCreate, TradeFillWrite
from app.services.trade_computation import FillComputationResult, compute_trade_from_fills


@dataclass
class TradeValidationResult:
    normalized_ticker: str
    fills: list[TradeFillWrite]
    fill_summary: FillComputationResult | None


def _raise_trade_validation_error(errors: list[str]) -> None:
    raise HTTPException(
        status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
        detail=" ".join(errors),
    )


def validate_trade_payload(payload: TradeCreate) -> TradeValidationResult:
    errors: list[str] = []
    normalized_ticker = payload.ticker.strip().upper()
    if not normalized_ticker:
        errors.append("Ticker is required.")

    if not isfinite(payload.entry_price) or payload.entry_price <= 0:
        errors.append("Entry price must be greater than zero.")

    if not isfinite(payload.exit_price) or payload.exit_price <= 0:
        errors.append("Exit price must be greater than zero.")

    if payload.quantity <= 0:
        errors.append("Quantity must be a whole number greater than zero.")

    if payload.setup_id is not None and payload.setup_id <= 0:
        errors.append("Setup must be a positive id when provided.")

    if payload.emotion_id is not None and payload.emotion_id <= 0:
        errors.append("Emotion must be a positive id when provided.")

    computed_pnl = (
        payload.pnl
        if payload.pnl is not None
        else payload.exit_price - payload.entry_price
    )
    if not isfinite(computed_pnl):
        errors.append("Computed PnL must be finite.")

    computed_total_pnl_usd = computed_pnl * payload.quantity * Trade.DEFAULT_CONTRACT_MULTIPLIER
    if not isfinite(computed_total_pnl_usd):
        errors.append("Total PnL must be finite.")

    use_fills = payload.use_fills or payload.fills is not None
    if not use_fills:
        if errors:
            _raise_trade_validation_error(errors)
        return TradeValidationResult(
            normalized_ticker=normalized_ticker,
            fills=[],
            fill_summary=None,
        )

    fills = payload.fills or []
    if not fills:
        errors.append("Add at least one BUY fill and one SELL fill.")
        _raise_trade_validation_error(errors)

    buy_fills = [fill for fill in fills if fill.side == "BUY"]
    sell_fills = [fill for fill in fills if fill.side == "SELL"]
    if not buy_fills or not sell_fills:
        errors.append("Add at least one BUY fill and one SELL fill.")

    total_buy_qty = sum(fill.quantity for fill in buy_fills)
    total_sell_qty = sum(fill.quantity for fill in sell_fills)
    if buy_fills and sell_fills and total_buy_qty != total_sell_qty:
        errors.append(
            "Open positions are not supported yet. Total BUY quantity "
            "must equal total SELL quantity."
        )

    if errors:
        deduped_errors: list[str] = []
        for error in errors:
            if error not in deduped_errors:
                deduped_errors.append(error)
        _raise_trade_validation_error(deduped_errors)

    fill_summary = compute_trade_from_fills(fills)
    if fill_summary.total_entry_qty <= 0:
        errors.append("Avg Entry must be a finite value greater than zero.")
    if fill_summary.total_exit_qty <= 0:
        errors.append("Avg Exit must be a finite value greater than zero.")
    if not isfinite(fill_summary.avg_entry_price) or fill_summary.avg_entry_price <= 0:
        errors.append("Avg Entry must be a finite value greater than zero.")
    if not isfinite(fill_summary.avg_exit_price) or fill_summary.avg_exit_price <= 0:
        errors.append("Avg Exit must be a finite value greater than zero.")
    if fill_summary.matched_qty <= 0:
        errors.append("Total Quantity must be a whole number greater than zero.")
    if not isfinite(fill_summary.realized_pnl_usd):
        errors.append("Total PnL must be finite.")

    if errors:
        deduped_errors: list[str] = []
        for error in errors:
            if error not in deduped_errors:
                deduped_errors.append(error)
        _raise_trade_validation_error(deduped_errors)

    return TradeValidationResult(
        normalized_ticker=normalized_ticker,
        fills=fills,
        fill_summary=fill_summary,
    )
